terminal/tcell/event.go

```diff
@@ -0,0 +1,192 @@
+	tcell.KeyCtrlB:          keyboard.KeyCtrlB,
+	tcell.KeyCtrlC:          keyboard.KeyCtrlC,
+	tcell.KeyCtrlD:          keyboard.KeyCtrlD,
+	tcell.KeyCtrlE:          keyboard.KeyCtrlE,
+	tcell.KeyCtrlF:          keyboard.KeyCtrlF,
+	tcell.KeyCtrlG:          keyboard.KeyCtrlG,
+	tcell.KeyCtrlJ:          keyboard.KeyCtrlJ,
+	tcell.KeyCtrlK:          keyboard.KeyCtrlK,
+	tcell.KeyCtrlL:          keyboard.KeyCtrlL,
+	tcell.KeyCtrlN:          keyboard.KeyCtrlN,
+	tcell.KeyCtrlO:          keyboard.KeyCtrlO,
+	tcell.KeyCtrlP:          keyboard.KeyCtrlP,
+	tcell.KeyCtrlQ:          keyboard.KeyCtrlQ,
+	tcell.KeyCtrlR:          keyboard.KeyCtrlR,
+	tcell.KeyCtrlS:          keyboard.KeyCtrlS,
+	tcell.KeyCtrlT:          keyboard.KeyCtrlT,
+	tcell.KeyCtrlU:          keyboard.KeyCtrlU,
+	tcell.KeyCtrlV:          keyboard.KeyCtrlV,
+	tcell.KeyCtrlW:          keyboard.KeyCtrlW,
+	tcell.KeyCtrlX:          keyboard.KeyCtrlX,
+	tcell.KeyCtrlY:          keyboard.KeyCtrlY,
+	tcell.KeyCtrlZ:          keyboard.KeyCtrlZ,
+	tcell.KeyBackspace:      keyboard.KeyBackspace,
+	tcell.KeyTab:            keyboard.KeyTab,
+	tcell.KeyEscape:         keyboard.KeyEsc,
+	tcell.KeyCtrlBackslash:  keyboard.KeyCtrlBackslash,
+	tcell.KeyCtrlRightSq:    keyboard.KeyCtrlRsqBracket,
+	tcell.KeyCtrlUnderscore: keyboard.KeyCtrlUnderscore,
+	tcell.KeyBackspace2:     keyboard.KeyBackspace2,
+	tcell.KeyCtrlSpace:      keyboard.KeyCtrlSpace,
+}
+
+// convKey converts a tcell keyboard event to the termdash format.
+func convKey(event *tcell.EventKey) terminalapi.Event {
+	tcellKey := event.Key()
+
+	if tcellKey == tcell.KeyRune {
+		ch := event.Rune()
+		return &terminalapi.Keyboard{
+			Key: keyboard.Key(ch),
+		}
+	}
+
+	k, ok := tcellToTd[tcellKey]
+	if !ok {
+		return terminalapi.NewErrorf("unknown keyboard key '%v' in a keyboard event", tcellKey)
+	}
+
+	return &terminalapi.Keyboard{
+		Key: k,
+	}
+}
+
+// convMouse converts a tcell mouse event to the termdash format.
+func convMouse(event *tcell.EventMouse) terminalapi.Event {
+	var button mouse.Button
+	x, y := event.Position()
+
+	tcellBtn := event.Buttons()
+
+	// tcell uses signed int16 for button masks, and negative values are invalid
+	if tcellBtn < 0 {
+		return terminalapi.NewErrorf("unknown mouse key %v in a mouse event", tcellBtn)
+	}
+
+	// Get wheel events
+	if tcellBtn&tcell.WheelUp != 0 {
+		button = mouse.ButtonWheelUp
+	} else if tcellBtn&tcell.WheelDown != 0 {
+		button = mouse.ButtonWheelDown
+	}
+
+	// Return wheel event if found
+	if button > 0 {
+		return &terminalapi.Mouse{
+			Position: image.Point{X: x, Y: y},
+			Button:   button,
+		}
+	}
+
+	// Get only button events, not wheel events
+	tcellBtn &= tcell.ButtonMask(0xff)
+	switch tcellBtn = event.Buttons(); tcellBtn {
+	case tcell.ButtonNone:
+		button = mouse.ButtonRelease
+	case tcell.Button1:
+		button = mouse.ButtonLeft
+	case tcell.Button2:
+		button = mouse.ButtonRight
+	case tcell.Button3:
+		button = mouse.ButtonMiddle
+	default:
```

Comment:
While this is according to what we discussed, the way it is implemented now the conversion layer would still forward an empty terminalapi.Mouse{} event with the default button and some x and y coordinates. This may confuse the upstream. A better approach might be to completely drop the event.

We might be able to accomplish that by modifying this function's signature so that it returns (terminalapi.Event, bool) and documenting that the bool is true iff we received an event that is valid for termdash. The caller then can check the bool and only forward the event if it is true.

Reply:
What would we do for callers of `convMouse()` then?

The only and immediate caller of `convMouse()` is `func toTermdashEvents(...) []terminalapi.Event` which returns a slice of `terminalapi.Event`s

If we change the signature of `convMouse()` to return -> `(terminalapi.Event, bool)`, how do we handle it from the toTermdashEvents function?

Reply:
For example, suppose that in https://github.com/mum4k/termdash/pull/224/files#diff-740d1ea120817d115c83619bb666a475R179 the code is changed to

```go
	case *tcell.EventMouse:
		mouseEvent, termdashOk := convMouse(event)
		...
		return []terminalapi.Event{mouseEvent}
```
and termdashOk is a boolean where convMouse returns true for all valid termdash event cases (and false for that default: case, e.g. where multiple mouse buttons are pressed at the same time)

What should be done if termdashOk == false? Since there needs to be a default case to match with the signature of `toTermdashEvents` returning `[]terminalapi.Event`. We can't have a `terminalapi.NewErrorf` event since that would crash the application.

Reply:
There is no direct connection between `toTermdashEvents()` and the main processing loop of Termdash. In fact they run in separate goroutines and there is an unbound queue in between them. The main processing loop of termdash receives events by periodically calling the `Event()` method on the terminal API:

https://github.com/mum4k/termdash/blob/68bf0566e6ec8effaf6e83400ef0046347a5f5f1/termdash.go#L309-L310

Because it is written this way, the main processing loop won't mind if events aren't available on every call, it will simply retry later (it is a bit more complicated, the main processing loop actually gets notified when the next event is available, but that isn't relevant here). Therefore all we need to do is simply avoid push the unsupported event onto the unbound queue.

Btw it just occurred to me that we may not need any special treatment or a boolean in the function signature, we can simply return a nil from `convMouse()` directly and look for that inside `toTermdashEvents()`.

And when `convMouse()` returns nil, `toTermdashEvents()` can also return nil (which is equivalent to an empty slice in Go, i.e. a func that returns the []foo type can return nil). The result of that should be that `pollEvents()` will get no events and thus push no events onto the unbound queue.

Please double check this logic, I may be missing something, but it does seem to check out.